File: frontend/src/app/run/runReadiness.ts
import { i18n } from '@/i18n'
import type { WorkflowStartRunView } from '@/app/transport/workflow'

export interface RunReadinessLike {
  state: string
  code?: string
  slot?: string
  nodeId?: string
  requirementId?: string
}

export type RunStartOutcome =
  | { state: 'started'; runId: string }
  | {
      state:
        | 'workflow-invalid'
        | 'target-required'
        | 'credential-required'
        | 'environment-unavailable'
        | 'not-started'
        | 'failed'
      code?: string
      slot?: string
      nodeId?: string
      requirementId?: string
    }

export function runStartOutcome(started: WorkflowStartRunView): RunStartOutcome {
  if (started.run) return { state: 'started', runId: started.run.runId }
  const readiness = started.readiness
  if (readiness && readiness.state !== 'started' && readiness.state !== 'failed') {
    return readinessOutcome(readiness)
  }
  if (readiness?.state === 'failed') {
    return readinessOutcome(readiness)
  }
  const diagnostic = started.diagnostics.find((item) => item.severity === 'error')
  if (diagnostic) {
    return { state: 'workflow-invalid', code: diagnostic.code, nodeId: diagnostic.nodeId }
  }
  return { state: 'not-started' }
}

export function readinessOutcome(
  readiness?: RunReadinessLike | null,
): Exclude<RunStartOutcome, { state: 'started' }> {
  if (!readiness) return { state: 'not-started' }
  return {
    state: normalizeReadinessState(readiness.state),
    code: readiness.code,
    slot: readiness.slot,
    nodeId: readiness.nodeId,
    requirementId: readiness.requirementId,
  }
}

export function runReadinessMessage(
  outcome: Exclude<RunStartOutcome, { state: 'started' }>,
): string {
  const t = i18n.global.t
  const te = i18n.global.te
  if (
    outcome.requirementId === 'model' &&
    (outcome.code === 'admission.target_unavailable' ||
      outcome.code === 'admission.target_ambiguous')
  ) {
    return t('workflow.run_readiness.ai_model_unavailable', { slot: outcome.slot ?? '' })
  }
  if (
    outcome.requirementId === 'model' &&
    (outcome.code === 'admission.credential_unavailable' ||
      outcome.code === 'admission.credential_ambiguous')
  ) {
    return t('workflow.run_readiness.ai_credential_unavailable', { slot: outcome.slot ?? '' })
  }
  if (outcome.code && te(`error.${outcome.code}`)) {
    return t(`error.${outcome.code}`, {
      slot: outcome.slot ?? '',
      nodeId: outcome.nodeId ?? '',
    })
  }
  return t(`workflow.run_readiness.${outcome.state}`)
}

function normalizeReadinessState(
  state: string,
): Exclude<RunStartOutcome, { state: 'started' }>['state'] {
  switch (state) {
    case 'workflow-invalid':
    case 'target-required':
    case 'credential-required':
    case 'environment-unavailable':
    case 'failed':
      return state
    default:
      return 'not-started'
  }
}

File: internal/application/run_readiness.go
package application

import (
	"errors"
	"strings"

	"github.com/yottaapp/yotta/internal/admission"
	"github.com/yottaapp/yotta/internal/workflow/schema"
)

type RunReadinessState string

const (
	RunReadinessStarted                RunReadinessState = "started"
	RunReadinessWorkflowInvalid        RunReadinessState = "workflow-invalid"
	RunReadinessTargetRequired         RunReadinessState = "target-required"
	RunReadinessCredentialRequired     RunReadinessState = "credential-required"
	RunReadinessEnvironmentUnavailable RunReadinessState = "environment-unavailable"
	RunReadinessNotStarted             RunReadinessState = "not-started"
	RunReadinessFailed                 RunReadinessState = "failed"
)

// RunReadiness is the application-level interpretation of a StartRun result.
// Every caller uses this classifier so a compiler diagnostic or admission
// requirement cannot be mistaken for a successfully queued Run.
type RunReadiness struct {
	State         RunReadinessState
	Code          string
	GraphID       string
	NodeID        string
	FromNodeID    string
	FromPortID    string
	ToNodeID      string
	ToPortID      string
	RequirementID string
	Slot          string
}

func ClassifyRunStart(result StartRunResult, startErr error) RunReadiness {
	if result.Record.Valid() {
		return RunReadiness{State: RunReadinessStarted}
	}
	for _, diagnostic := range result.Diagnostics {
		if diagnostic.Severity == schema.SeverityError {
			return RunReadiness{
				State:      RunReadinessWorkflowInvalid,
				Code:       diagnostic.Code,
				GraphID:    strings.Join(diagnostic.GraphPath, "/"),
				NodeID:     diagnostic.NodeID,
				FromNodeID: diagnosticStringParam(diagnostic, "fromNodeId"),
				FromPortID: diagnosticStringParam(diagnostic, "fromPortId"),
				ToNodeID:   diagnosticStringParam(diagnostic, "toNodeId"),
				ToPortID:   diagnosticStringParam(diagnostic, "toPortId"),
			}
		}
	}
	var admissionErr *admission.Error
	if errors.As(startErr, &admissionErr) {
		state := RunReadinessFailed
		switch admissionErr.Code {
		case admission.CodeTargetUnavailable, admission.CodeTargetAmbiguous:
			state = RunReadinessTargetRequired
		case admission.CodeCredentialUnavailable, admission.CodeCredentialAmbiguous:
			state = RunReadinessCredentialRequired
		case admission.CodeProviderIncompatible, admission.CodeUnsupportedHost:
			state = RunReadinessEnvironmentUnavailable
		}
		return RunReadiness{
			State: state, Code: admissionErr.Code, GraphID: admissionErr.GraphID,
			NodeID: admissionErr.NodeID, RequirementID: admissionErr.RequirementID, Slot: admissionErr.Slot,
		}
	}
	if startErr != nil {
		return RunReadiness{State: RunReadinessFailed}
	}
	return RunReadiness{State: RunReadinessNotStarted}
}

func diagnosticStringParam(diagnostic schema.Diagnostic, key string) string {
	value, _ := diagnostic.Params[key].(string)
	return value
}

func (r RunReadiness) UserFixable() bool {
	switch r.State {
	case RunReadinessWorkflowInvalid, RunReadinessTargetRequired, RunReadinessCredentialRequired,
		RunReadinessEnvironmentUnavailable, RunReadinessNotStarted:
		return true
	default:
		return false
	}
}
